verifier/trustpolicy/blob.go

```diff
@@ -0,0 +1,133 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trustpolicy
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/notaryproject/notation-go/dir"
+	set "github.com/notaryproject/notation-go/internal/container"
+	"github.com/notaryproject/notation-go/internal/slices"
+)
+
+// BlobDocument represents a trustpolicy.blob.json document
+type BlobDocument struct {
+	// Version of the policy document
+	Version string `json:"version"`
+
+	// BlobTrustPolicies include each policy statement
+	BlobTrustPolicies []BlobTrustPolicy `json:"trustPolicies"`
+}
+
+// BlobTrustPolicy represents a policy statement in the blob policy document
+type BlobTrustPolicy struct {
+	// Name of the policy statement
+	Name string `json:"name"`
+
+	// SignatureVerification setting for this policy statement
+	SignatureVerification SignatureVerification `json:"signatureVerification"`
+
+	// TrustStores this policy statement uses
+	TrustStores []string `json:"trustStores"`
+
+	// TrustedIdentities this policy statement pins
+	TrustedIdentities []string `json:"trustedIdentities"`
+
+	// GlobalPolicy defines if policy statement is global or not
+	GlobalPolicy bool `json:"globalPolicy,omitempty"`
+}
+
+// LoadBlobDocument loads a trust policy document from a local file system
+func LoadBlobDocument() (*BlobDocument, error) {
+	var doc BlobDocument
+	err := getDocument(dir.PathBlobTrustPolicy, &doc)
+	return &doc, err
+}
+
+// Validate validates a policy document according to its version's rule set.
+// if any rule is violated, returns an error
+func (policyDoc *BlobDocument) Validate() error {
+	// sanity check
+	if policyDoc == nil {
+		return errors.New("blob trust policy document cannot be nil")
+	}
+
+	// Validate Version
+	if policyDoc.Version == "" {
+		return errors.New("blob trust policy has empty version, version must be specified")
+	}
+	if !slices.Contains(supportedPolicyVersions, policyDoc.Version) {
+		return fmt.Errorf("blob trust policy document uses unsupported version %q", policyDoc.Version)
+	}
+
+	// Validate the policy according to 1.0 rules
+	if len(policyDoc.BlobTrustPolicies) == 0 {
+		return errors.New("blob trust policy document can not have zero trust policy statements")
+	}
+
+	policyNames := set.New[string]()
+	foundGlobalPolicy := false
+	for _, statement := range policyDoc.BlobTrustPolicies {
+		// Verify unique policy statement names across the policy document
+		if policyNames.Contains(statement.Name) {
+			return fmt.Errorf("multiple blob trust policy statements use the same name %q, statement names must be unique", statement.Name)
+		}
+		policyNames.Add(statement.Name)
```

Comment:
You should validate the `statement.Name` exists before making use of it for uniqueness check.

Reply:
good catch, moved `policyNames.Add` to end of for loop.